File: hearthline-alexa-mcp/src/mcp-server.mjs
import http from 'node:http';
import { randomUUID } from 'node:crypto';
import { APP_MIME, APP_URI, dashboardHtml } from './app-resource.mjs';
import { callTool, listTools } from './tools.mjs';

export const PROTOCOL_VERSION = '2025-11-25';
const MAX_BODY = 1_000_000;

function jsonRpcResult(id, result) { return { jsonrpc: '2.0', id, result }; }
function jsonRpcError(id, code, message, data) { return { jsonrpc: '2.0', id: id ?? null, error: { code, message, ...(data === undefined ? {} : { data }) } }; }
function validRequestId(id) { return typeof id === 'string' || (typeof id === 'number' && Number.isInteger(id)); }

function isAllowedOrigin(origin, allowedOrigins) {
  if (!origin) return true;
  if (allowedOrigins.has(origin)) return true;
  try { const url = new URL(origin); return (url.hostname === '127.0.0.1' || url.hostname === 'localhost') && (url.protocol === 'http:' || url.protocol === 'https:'); }
  catch { return false; }
}

async function readJson(req) {
  let bytes = 0; const chunks = [];
  for await (const chunk of req) { bytes += chunk.length; if (bytes > MAX_BODY) throw Object.assign(new Error('request body too large'), { status: 413 }); chunks.push(chunk); }
  if (bytes === 0) throw Object.assign(new Error('empty request body'), { status: 400 });
  try { return JSON.parse(Buffer.concat(chunks).toString('utf8')); } catch { throw Object.assign(new Error('invalid JSON'), { status: 400 }); }
}

export function createMcpHttpServer({ orchestrator, allowedOrigins = [], logger = console, sessionTtlMs = 30 * 60_000 }) {
  if (!Number.isFinite(sessionTtlMs) || sessionTtlMs <= 0) throw new Error('sessionTtlMs must be finite and > 0');
  const sessions = new Map(); const allow = new Set(allowedOrigins);
  const server = http.createServer(async (req, res) => {
    try {
      pruneExpiredSessions(sessions, Date.now(), sessionTtlMs);
      const url = new URL(req.url, 'http://localhost');
      if (url.pathname === '/health' && req.method === 'GET') return sendJson(res, 200, { ok: true, protocolVersion: PROTOCOL_VERSION, sessions: sessions.size });
      if (url.pathname !== '/mcp') return sendJson(res, 404, { error: 'not found' });
      if (!isAllowedOrigin(req.headers.origin, allow)) return sendJson(res, 403, { error: 'origin not allowed' });
      if (req.method === 'GET') { res.writeHead(405, { Allow: 'POST, DELETE', 'Cache-Control': 'no-store' }); return res.end(); }
      if (req.method === 'DELETE') {
        const sessionId = req.headers['mcp-session-id'];
        if (!sessionId) return sendRpc(res, 400, jsonRpcError(null, -32001, 'MCP-Session-Id is required'));
        if (!sessions.has(sessionId)) return sendRpc(res, 404, jsonRpcError(null, -32001, 'Unknown or expired MCP session'));
        const protocol = req.headers['mcp-protocol-version'];
        if (protocol !== PROTOCOL_VERSION) return sendRpc(res, 400, jsonRpcError(null, -32002, `MCP-Protocol-Version must be ${PROTOCOL_VERSION}`), sessionId);
        sessions.delete(sessionId);
        res.writeHead(204, { 'Cache-Control': 'no-store' });
        return res.end();
      }
      if (req.method !== 'POST') { res.writeHead(405, { Allow: 'GET, POST, DELETE' }); return res.end(); }
      const accept = String(req.headers.accept ?? '');
      if (!accept.includes('application/json') || !accept.includes('text/event-stream')) return sendRpc(res, 406, jsonRpcError(null, -32000, 'Accept must include application/json and text/event-stream'));
      const message = await readJson(req);
      if (!message || message.jsonrpc !== '2.0' || typeof message.method !== 'string') return sendRpc(res, 400, jsonRpcError(message?.id, -32600, 'Invalid Request'));
      if (Object.prototype.hasOwnProperty.call(message, 'id') && !validRequestId(message.id)) return sendRpc(res, 400, jsonRpcError(null, -32600, 'Invalid Request: id must be a string or integer'));
      const isInitialize = message.method === 'initialize';
      if (isInitialize && message.id === undefined) return sendRpc(res, 400, jsonRpcError(null, -32600, 'initialize must be a JSON-RPC request with an id'));
      let sessionId = req.headers['mcp-session-id'];
      if (isInitialize) {
        const requested = message.params?.protocolVersion;
        if (typeof requested !== 'string' || requested.length === 0) return sendRpc(res, 200, jsonRpcError(message.id, -32602, 'initialize requires protocolVersion'));
        sessionId = randomUUID(); sessions.set(sessionId, { createdAt: Date.now(), lastSeenAt: Date.now(), initialized: false, toolCalls: [] });
      }
      else {
        if (!sessionId) return sendRpc(res, 400, jsonRpcError(message.id, -32001, 'MCP-Session-Id is required'));
        if (!sessions.has(sessionId)) return sendRpc(res, 404, jsonRpcError(message.id, -32001, 'Unknown or expired MCP session'));
        const protocol = req.headers['mcp-protocol-version'];
        if (protocol !== PROTOCOL_VERSION) return sendRpc(res, 400, jsonRpcError(message.id, -32002, `MCP-Protocol-Version must be ${PROTOCOL_VERSION}`), sessionId);
        sessions.get(sessionId).lastSeenAt = Date.now();
      }
      if (message.id === undefined) { if (message.method === 'notifications/initialized' && sessions.has(sessionId)) sessions.get(sessionId).initialized = true; res.writeHead(202, { 'MCP-Session-Id': sessionId, 'Cache-Control': 'no-store' }); return res.end(); }
      if (!isInitialize && message.method !== 'ping' && !sessions.get(sessionId).initialized) return sendRpc(res, 400, jsonRpcError(message.id, -32003, 'session not initialized; send notifications/initialized first'), sessionId);
      let result;
      try { result = await dispatch(message.method, message.params ?? {}, orchestrator, sessions, sessionId); }
      catch (error) { logger.error?.('mcp method error', { method: message.method, error: error?.message }); return sendRpc(res, 200, jsonRpcError(message.id, error?.jsonRpcCode ?? -32010, error?.message ?? 'method execution failed'), sessionId); }
      return sendRpc(res, 200, jsonRpcResult(message.id, result), sessionId);
    } catch (error) { const status = Number(error?.status ?? 500); logger.error?.('mcp transport error', { error: error?.message, status }); return sendRpc(res, status, jsonRpcError(null, status === 400 ? -32700 : -32000, error?.message ?? 'transport failure')); }
  });
  return server;
}

async function dispatch(method, params, orchestrator, sessions, sessionId) {
  switch (method) {
    case 'initialize': return { protocolVersion: PROTOCOL_VERSION, capabilities: { tools: { listChanged: false }, resources: { subscribe: false, listChanged: false } }, serverInfo: { name: 'Hearthline Household Mission MCP', version: '0.1.0' }, instructions: 'Use Hearthline to build stateful household missions. External-commit actions require explicit approval, then idempotent execution. Shopping demo actions never place purchases.' };
    case 'ping': return {};
    case 'tools/list': return { tools: listTools() };
    case 'tools/call': {
      if (typeof params?.name !== 'string') throw protocolError(-32602, 'tools/call requires name');
      if (!listTools().some((tool) => tool.name === params.name)) throw protocolError(-32602, `unknown tool: ${params.name}`);
      enforceToolRateLimit(sessions.get(sessionId));
      try { return { ...(await callTool(orchestrator, params.name, params.arguments ?? {})), isError: false }; }
      catch (error) { return { content: [{ type: 'text', text: `Hearthline rejected the tool call: ${String(error?.message ?? 'execution failed').slice(0, 1000)}` }], isError: true }; }
    }
    case 'resources/list': return { resources: [{ uri: APP_URI, name: 'Hearthline Mission Dashboard', description: 'Interactive mission status and approval dashboard', mimeType: APP_MIME }] };
    case 'resources/read': if (params?.uri !== APP_URI) throw new Error('resource not found'); else return { contents: [{ uri: APP_URI, mimeType: APP_MIME, text: dashboardHtml(), _meta: { ui: { prefersBorder: true } } }] };
    default: throw protocolError(-32601, `method not found: ${method}`);
  }
}
function protocolError(code, message) { return Object.assign(new Error(message), { jsonRpcCode: code }); }
function pruneExpiredSessions(sessions, now, ttlMs) { for (const [id, session] of sessions) if (now - session.lastSeenAt > ttlMs) sessions.delete(id); }
function enforceToolRateLimit(session, now = Date.now()) { const windowStart = now - 60_000; session.toolCalls = session.toolCalls.filter((time) => time >= windowStart); if (session.toolCalls.length >= 120) throw new Error('tool rate limit exceeded; retry after the current one-minute window'); session.toolCalls.push(now); }
function sendRpc(res, status, body, sessionId) { const headers = { 'Content-Type': 'application/json; charset=utf-8', 'Cache-Control': 'no-store' }; if (sessionId) headers['MCP-Session-Id'] = sessionId; res.writeHead(status, headers); res.end(JSON.stringify(body)); }
function sendJson(res, status, body) { res.writeHead(status, { 'Content-Type': 'application/json; charset=utf-8', 'Cache-Control': 'no-store' }); res.end(JSON.stringify(body)); }

File: amazon-alexa-homeops-relay/mcp_server.py
from __future__ import annotations

import json
import secrets
from http import HTTPStatus
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import Any
from urllib.parse import urlsplit

from core import HomeOpsError, HomeOpsStore, canonical_json

PROTOCOL_VERSION = "2025-11-25"
SERVER_NAME = "homeops-relay"
SERVER_VERSION = "1.0.0"
MAX_BODY = 64 * 1024


def _pairs(items):
    out = {}
    for key, value in items:
        if key in out:
            raise HomeOpsError(f"duplicate JSON key: {key}")
        out[key] = value
    return out


def _bad_number(value):
    raise HomeOpsError(f"non-integer/non-finite JSON number forbidden: {value}")


def strict_json(raw: bytes) -> dict[str, Any]:
    if raw.startswith(b"\xef\xbb\xbf"):
        raise HomeOpsError("JSON BOM forbidden")
    value = json.loads(
        raw.decode("utf-8"),
        object_pairs_hook=_pairs,
        parse_float=_bad_number,
        parse_constant=_bad_number,
    )
    if not isinstance(value, dict):
        raise HomeOpsError("JSON-RPC request object required")
    return value

TOOL_SCHEMAS: dict[str, dict[str, Any]] = {
    "homeops.create_issue": {
        "description": "Create a household maintenance issue in the local evidence ledger.",
        "inputSchema": {"type": "object", "additionalProperties": False, "required": ["issue_id", "title", "description", "priority"], "properties": {"issue_id": {"type": "string"}, "title": {"type": "string"}, "description": {"type": "string"}, "priority": {"type": "string", "enum": ["LOW", "MEDIUM", "HIGH", "URGENT"]}}},
    },
    "homeops.add_evidence": {
        "description": "Attach bounded evidence to an issue; no external fetch is performed.",
        "inputSchema": {"type": "object", "additionalProperties": False, "required": ["issue_id", "evidence_id", "kind", "summary", "source"], "properties": {k: {"type": "string"} for k in ["issue_id", "evidence_id", "kind", "summary", "source"]}},
    },
    "homeops.add_quote": {
        "description": "Record a vendor quote as evidence without purchasing or contacting the vendor.",
        "inputSchema": {"type": "object", "additionalProperties": False, "required": ["issue_id", "quote_id", "vendor", "amount_cents", "currency", "scope"], "properties": {"issue_id": {"type": "string"}, "quote_id": {"type": "string"}, "vendor": {"type": "string"}, "amount_cents": {"type": "integer", "minimum": 0}, "currency": {"type": "string"}, "scope": {"type": "string"}}},
    },
    "homeops.propose_plan": {
        "description": "Create an evidence-bound proposal. The result is proposal-only and carries no execution authority.",
        "inputSchema": {"type": "object", "additionalProperties": False, "required": ["issue_id", "plan_id", "summary", "steps"], "properties": {"issue_id": {"type": "string"}, "plan_id": {"type": "string"}, "summary": {"type": "string"}, "steps": {"type": "array", "minItems": 1, "maxItems": 16, "items": {"type": "string"}}}},
    },
    "homeops.review_plan": {
        "description": "Record an explicit human plan approval or rejection.",
        "inputSchema": {"type": "object", "additionalProperties": False, "required": ["plan_id", "decision", "reviewer", "note"], "properties": {"plan_id": {"type": "string"}, "decision": {"type": "string", "enum": ["APPROVE", "REJECT"]}, "reviewer": {"type": "string"}, "note": {"type": "string"}}},
    },
    "homeops.request_side_effect": {
        "description": "Compile an approved side-effect request for a separate executor. This server never executes the action.",
        "inputSchema": {"type": "object", "additionalProperties": False, "required": ["plan_id", "action_id", "kind", "summary"], "properties": {"plan_id": {"type": "string"}, "action_id": {"type": "string"}, "kind": {"type": "string", "enum": ["SEND_MESSAGE", "PLACE_ORDER", "SCHEDULE_VISIT", "CHANGE_DEVICE_STATE"]}, "summary": {"type": "string"}}},
    },
    "homeops.get_issue": {
        "description": "Read one issue and its evidence, quotes, and plans.",
        "inputSchema": {"type": "object", "additionalProperties": False, "required": ["issue_id"], "properties": {"issue_id": {"type": "string"}}},
    },
    "homeops.verify_event_chain": {
        "description": "Verify the local content-addressed event chain.",
        "inputSchema": {"type": "object", "additionalProperties": False, "properties": {}},
    },
}


class MCPApplication:
    def __init__(self) -> None:
        self.store = HomeOpsStore()
        self.sessions: dict[str, bool] = {}

    def initialize(self, params: Any) -> tuple[dict[str, Any], str]:
        if not isinstance(params, dict):
            raise HomeOpsError("initialize params must be object")
        if params.get("protocolVersion") != PROTOCOL_VERSION:
            raise HomeOpsError("protocolVersion 2025-11-25 required")
        if not isinstance(params.get("capabilities", {}), dict) or not isinstance(params.get("clientInfo", {}), dict):
            raise HomeOpsError("invalid initialize capabilities/clientInfo")
        session_id = secrets.token_urlsafe(24)
        self.sessions[session_id] = False
        return {
            "protocolVersion": PROTOCOL_VERSION,
            "capabilities": {"tools": {"listChanged": False}},
            "serverInfo": {"name": SERVER_NAME, "version": SERVER_VERSION},
            "instructions": "HomeOps Relay is local-first and proposal-only. Side-effect tools return requests for a separate human-controlled executor; they never send, purchase, schedule, or actuate.",
        }, session_id

    def dispatch(self, method: str, params: Any) -> Any:
        if method == "ping":
            return {}
        if method == "tools/list":
            return {"tools": [{"name": name, **schema} for name, schema in sorted(TOOL_SCHEMAS.items())]}
        if method == "tools/call":
            if not isinstance(params, dict) or set(params) - {"name", "arguments", "_meta"}:
                raise HomeOpsError("tools/call: invalid params")
            name = params.get("name")
            args = params.get("arguments", {})
            if name not in TOOL_SCHEMAS or not isinstance(args, dict):
                raise HomeOpsError("tools/call: unknown tool or invalid arguments")
            fn_name = name.split(".", 1)[1]
            fn = getattr(self.store, fn_name)
            result = fn(args) if fn_name != "verify_event_chain" else fn()
            return {
                "content": [{"type": "text", "text": json.dumps(result, sort_keys=True, separators=(",", ":"))}],
                "structuredContent": result,
                "isError": False,
            }
        raise KeyError(method)


class MCPHandler(BaseHTTPRequestHandler):
    app = MCPApplication()
    server_version = "HomeOpsRelay/1.0"

    def log_message(self, fmt: str, *args: Any) -> None:
        return

    def _headers_ok(self) -> bool:
        host = self.headers.get("Host", "")
        if not host:
            return False
        origin = self.headers.get("Origin")
        if origin:
            parsed = urlsplit(origin)
            origin_host = parsed.netloc.lower()
            if parsed.scheme not in {"http", "https"} or origin_host != host.lower():
                return False
        return True

    def _json(self, status: int, body: dict[str, Any] | None, *, session_id: str | None = None) -> None:
        raw = b"" if body is None else canonical_json(body)
        self.send_response(status)
        if body is not None:
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(raw)))
        if session_id:
            self.send_header("Mcp-Session-Id", session_id)
        self.end_headers()
        if raw:
            self.wfile.write(raw)

    def do_DELETE(self) -> None:
        if self.path != "/mcp" or not self._headers_ok():
            self._json(HTTPStatus.FORBIDDEN, {"error": "forbidden"})
            return
        session = self.headers.get("Mcp-Session-Id")
        if not session or session not in self.app.sessions:
            self._json(HTTPStatus.NOT_FOUND, {"error": "unknown session"})
            return
        del self.app.sessions[session]
        self._json(HTTPStatus.NO_CONTENT, None)

    def do_GET(self) -> None:
        if self.path == "/healthz":
            self._json(HTTPStatus.OK, {"ok": True, "protocol": PROTOCOL_VERSION, "execution_authority": False})
            return
        if self.path == "/mcp":
            self.send_response(HTTPStatus.METHOD_NOT_ALLOWED)
            self.send_header("Allow", "POST, DELETE")
            self.end_headers()
            return
        self._json(HTTPStatus.NOT_FOUND, {"error": "not found"})

    def do_POST(self) -> None:
        if self.path != "/mcp":
            self._json(HTTPStatus.NOT_FOUND, {"error": "not found"})
            return
        if not self._headers_ok():
            self._json(HTTPStatus.FORBIDDEN, {"error": "host/origin rejected"})
            return
        ctype = self.headers.get("Content-Type", "").split(";", 1)[0].strip().lower()
        accept = self.headers.get("Accept", "")
        if ctype != "application/json" or ("application/json" not in accept and "text/event-stream" not in accept):
            self._json(HTTPStatus.UNSUPPORTED_MEDIA_TYPE, {"error": "application/json request and JSON/SSE accept required"})
            return
        try:
            length = int(self.headers.get("Content-Length", "0"))
        except ValueError:
            self._json(HTTPStatus.BAD_REQUEST, {"error": "invalid content length"})
            return
        if length <= 0 or length > MAX_BODY:
            self._json(HTTPStatus.REQUEST_ENTITY_TOO_LARGE, {"error": "invalid body size"})
            return
        try:
            raw = self.rfile.read(length)
            obj = strict_json(raw)
            if obj.get("jsonrpc") != "2.0" or not isinstance(obj.get("method"), str) or set(obj) - {"jsonrpc", "id", "method", "params"}:
                raise HomeOpsError("invalid JSON-RPC request")
            method = obj["method"]
            is_notification = "id" not in obj
            if method == "initialize":
                if is_notification:
                    raise HomeOpsError("initialize requires id")
                result, session = self.app.initialize(obj.get("params", {}))
                self._json(HTTPStatus.OK, {"jsonrpc": "2.0", "id": obj["id"], "result": result}, session_id=session)
                return
            session = self.headers.get("Mcp-Session-Id")
            if not session or session not in self.app.sessions:
                self._json(HTTPStatus.NOT_FOUND, {"error": "unknown session"})
                return
            if self.headers.get("MCP-Protocol-Version") not in {None, "", PROTOCOL_VERSION}:
                self._json(HTTPStatus.BAD_REQUEST, {"error": "protocol version mismatch"})
                return
            if method == "notifications/initialized":
                if not is_notification:
                    self._json(HTTPStatus.OK, {"jsonrpc": "2.0", "id": obj["id"], "error": {"code": -32600, "message": "initialized must be a notification"}})
                    return
                self.app.sessions[session] = True
                self._json(HTTPStatus.ACCEPTED, None)
                return
            if not self.app.sessions[session]:
                if is_notification:
                    self._json(HTTPStatus.ACCEPTED, None)
                else:
                    self._json(HTTPStatus.OK, {"jsonrpc": "2.0", "id": obj["id"], "error": {"code": -32002, "message": "Server not initialized"}})
                return
            if is_notification:
                self._json(HTTPStatus.ACCEPTED, None)
                return
            try:
                result = self.app.dispatch(method, obj.get("params", {}))
                body = {"jsonrpc": "2.0", "id": obj["id"], "result": result}
            except KeyError:
                body = {"jsonrpc": "2.0", "id": obj["id"], "error": {"code": -32601, "message": "Method not found"}}
            except HomeOpsError as exc:
                if method == "tools/call":
                    body = {"jsonrpc": "2.0", "id": obj["id"], "result": {"content": [{"type": "text", "text": f"Error: {exc}"}], "isError": True}}
                else:
                    body = {"jsonrpc": "2.0", "id": obj["id"], "error": {"code": -32602, "message": str(exc)}}
            self._json(HTTPStatus.OK, body)
        except (UnicodeDecodeError, json.JSONDecodeError, HomeOpsError) as exc:
            self._json(HTTPStatus.BAD_REQUEST, {"jsonrpc": "2.0", "id": None, "error": {"code": -32700, "message": str(exc)}})


def serve(host: str = "127.0.0.1", port: int = 8787) -> None:
    httpd = ThreadingHTTPServer((host, port), MCPHandler)
    print(f"HomeOps Relay MCP {PROTOCOL_VERSION} on http://{host}:{port}/mcp")
    httpd.serve_forever()


if __name__ == "__main__":
    serve()
